spec.md

```diff
@@ -0,0 +1,1597 @@
+
+_Comment: notice difference in capitalization between the two cases!_
+
+## Redefine semantics
+
+_In each case, redefine acts like an undefine (which cannot be a no-op) and a define of the given axiom_
+
+**Redefine principles**: 
+
+1. Can only redefine one thing at a time
+2. We disallow redefining boolean properties:
+  * _Example 1_: a type can either exists or not. we cannot "redefine" it's existence, but only define or undefine it.
+  * _Example 2_: a type is either abstract or not. we can only define or undefine `@abstract`.
+
+### Type axioms
+
+**Case ENT**
+* cannot redefine `entity A`
+* `(entity) A sub B` redefines $`A < B`$
+
+**Case REL**
+* cannot redefine `relation A` 
+* `(relation) A sub B` redefines $`A < B`$
+* cannot redefine `(relation) A relates I`
+* `(relation) A relates I as J$` redefines $`I <_! J`$, ***requiring*** that either $`I <_! J' \neq J`$ or $`I`$ has no direct super-role
+* cannot redefine `(relation) A relates I[]`
+* `(relation) A relates I[] as J[]` redefines $`I <_! J`$, ***requiring*** that either $`I <_! J' \neq J`$ or $`I`$ has no direct super-role
+
+**Case ATT**
+* cannot redefine `attribute A`
+* `(attribute) A value V` redefines $`A < V`$
+* cannot redefine `(attribute) A sub B`
+
+**Case PLAYS**
+* cannot redefine `(type) A plays B:I`
+
+**Case OWNS**
+* cannot redefine `(type) A owns B`
+* cannot redefine `(type) A owns B[]`
+
+### Constraints
+
+_In each case, `redefine` redefines the postulated condition._
+
+**Case CARD**
+* `A relates I @card(n..m)`
+* `A plays B:I @card(n..m)`
+* `A owns B @card(n...m)`
+
+**Case CARD_LIST**
+* `A relates I[] @card(n..m)`
+* `A owns B[] @card(n...m)`
+
+**Case PLAYS_AS**
+* `A plays B as C`
+
+**Case OWNS_AS**
+* `A owns B as C`
+
+**Case UNIQUE**
+* cannot redefine `A owns B @unique`
+
+**Case KEY**
+* cannot redefine `A owns B @key`
+
+**Case SUBKEY**
+* cannot redefine `A owns B @subkey(<LABEL>)`
+
+**Case ABSTRACT**
+* cannot redefine `(type) B @abstract` 
+* cannot redefine `A plays B:I @abstract`
+* cannot redefine `A owns B @abstract` 
+* cannot redefine `A owns B[] @abstract` 
+* cannot redefine `B relates I @abstract`
+* cannot redefine `B relates I[] @abstract`
+
+**Case VALUES**
+* `A owns B @values(v1, v2)` 
+* `A owns B @regex(<EXPR>)` 
+* `A owns B @range(v1..v2)`
+* `A value B @values(v1, v2)` 
+* `A value B @regex(<EXPR>)` 
+* `A value B @range(v1..v2)`
+
+**Case DISTINCT**
+* cannot redefine `A owns B[] @distinct`
+* cannot redefine `B relates I[] @distinct`
+
+### Triggers
+
+_In each case, `redefine` redefines the triggered action._
+
+**Case DEP_DEL (CASCADE/INDEPEDENT)**
+* cannot redefine `(relation) B relates I @cascade`
+* cannot redefine `(relation) B @cascade`
+* cannot redefine `(attribute) B @independent`
+
+### Value types
+
+**Case PRIMITIVES**
+
+cannot redefine primitives
+
+**Case STRUCT**
+
+cannot redefine structs
+
+### Functions defs
+
+**Case STREAM_RET_FUN**
+
+cannot redefine stream-return functions.
```

Comment:
This will be very inconvient, e.g. I have a series of functions A - D, each calling the previous, and call D in my query. I discover a bug in A. With the current undefine and redefine semantics, I would need to undefine D, C, B, A, then define them A', B, C, D. This means that a bug in a function that is depended on throughout the codebase could lead to massive refactors. I don't see a reason why a function couldn't be redefined if its signature remains the same. Ditto below for single-return functions.